load_balancing: allow to disable shuffling replicas

piodul opened the pull request:
## Pre-review checklist

When token awareness is enabled in the default load balancing policy, it chooses a random replica to send the query to. This is supposed to balance the load however in some scenarios it might be more beneficial to always choose the same replica - for example, reads might benefit from better caching. This PR adds an option to disable this behavior, which will cause the load balancing policy to always choose nodes in the same order for each replica set.

This feature was inspired by similar functionality in the cpp driver.

Fixes: #634

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

wprzytula commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -323,11 +330,18 @@ impl DefaultPolicy {
 
-        let mut rng = thread_rng();
-        vec.shuffle(&mut rng);
+        if let Some(fixed) = self.fixed_shuffle_seed {
+            let mut gen = Pcg32::new(fixed as u64, 0);
+            vec.shuffle(&mut gen);
```
Is the cost of calling `Pcg32::new()` on each query negligible, or should we take it into account?

piodul replied:
It should be pretty cheap: https://docs.rs/rand_pcg/latest/src/rand_pcg/pcg64.rs.html#88-92. I can see one multiplication, two additions and some operations on bits.

piodul commented:
Accidentally marked as draft, un-drafting.

wprzytula commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -651,13 +681,37 @@ mod tests {
+                        assert_eq!(&got_set, expected_set);
+
+                        // Put the group into sets_of_groups
+                        sets_of_groups[group_id].insert(got_group.clone());
```
Isn't this `clone()` unnecessary?

piodul replied:
Yes, it is, I will get rid of it.

wprzytula commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -606,12 +606,21 @@ mod tests {
         enum ExpectedGroup {
             AnyOrder(HashSet<u16>),
+            Deterministic(HashSet<u16>),
         }
```
I like this design very much! The types rule.

wprzytula commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -984,6 +1009,31 @@ mod tests {
+                //                                      us  eu  eu  us  eu  eu  us
+                //                                      r2  r1  r1  r1  r2  r1  r1
+                expected_groups: ExpectedGroupsBuilder::new()
+                    .deterministic([A, C, G]) // pick + fallback local replicas
+                    .deterministic([F, D, E]) // remote replicas
```
How did you find out the output order for the given seed? Purely empirically?

piodul replied:
I don't assume that the nodes within the group will be in some particular order. I just go over every plan that is being checked and verify that they all have the same ordering of nodes within that group.

wprzytula replied:
Oh, right.

wprzytula commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -605,21 +605,21 @@ mod tests {
-            AnyOrder(HashSet<u16>),
+            NonDeterministic(HashSet<u16>),
             Deterministic(HashSet<u16>),
         }
```
So now we expect nondeterminism both for replicas and for non-replicas, which is a generalisation. Replicas are shuffled, whereas non-replicas are randomly rotated. If we want to consider both as nondeterministic, *dot*, then we can leave it as you propose; if, however, the distinction between them is meaningful, we should highlight that distinction at least in naming.

piodul replied:
I'd rather leave this as "non-deterministic". If we wanted to differentiate "shuffling" and "rotating" then we probably could provide different assertions for both types. I'd rather not do it in this PR.

piodul commented:
v2:
- Changed the `fixed_shuffle_seed` size from `u32` to `u64`
- Got rid of an unnecessary clone in `assert_proper_grouping_in_plans`
- Added `fixed_shuffle_seed` to the `impl Debug`
